File: apps/api/src/modules/generation/rendercv-renderer.ts
import { Injectable, Logger } from '@nestjs/common';
import { execFile } from 'child_process';
import { mkdir, writeFile } from 'fs/promises';
import * as path from 'path';
import { promisify } from 'util';
import { stringify as yamlStringify } from 'yaml';
import { RendercvMaster } from './rendercv-tailor';

const execFileAsync = promisify(execFile);

/**
 * Renders a (tailored) RenderCV master object to PDF by writing it back to YAML and
 * shelling out to the `rendercv` CLI. Sits alongside {@link HtmlPdfRenderer}: the
 * JSON-Resume path uses Handlebars + Puppeteer, this path uses the user's real
 * rendercv config so the Typst theme/design is preserved exactly.
 */
@Injectable()
export class RenderCvRenderer {
  private readonly logger = new Logger(RenderCvRenderer.name);

  private get outDir(): string {
    return process.env.DOCUMENTS_DIR ?? '/data/documents';
  }

  private get bin(): string {
    return process.env.RENDERCV_BIN ?? 'rendercv';
  }

  /**
   * @param master merged rendercv config (design/templates/locale intact)
   * @param baseName sanitized file stem, e.g. "acme-backend-engineer-v1"
   * @returns absolute path to the produced PDF
   */
  async render(master: RendercvMaster, baseName: string): Promise<{ yamlPath: string; pdfPath: string }> {
    await mkdir(this.outDir, { recursive: true });
    const yamlPath = path.join(this.outDir, `${baseName}.yaml`);
    const pdfPath = path.join(this.outDir, `${baseName}.pdf`);

    await writeFile(yamlPath, yamlStringify(master), 'utf8');

    // -pdf path is relative to the input file, so pass just the basename.
    const { stderr } = await execFileAsync(
      this.bin,
      ['render', yamlPath, '-o', this.outDir, '-pdf', `${baseName}.pdf`, '-nopng', '-nohtml', '-nomd'],
      { cwd: this.outDir, timeout: 120_000 },
    );
    if (stderr?.trim()) this.logger.debug(`rendercv stderr: ${stderr.trim()}`);

    return { yamlPath, pdfPath };
  }
}

File: apps/api-go/internal/generation/rendercv_renderer.go
package generation

import (
	"bytes"
	"context"
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"time"

	"gopkg.in/yaml.v3"
)

// RenderCvRenderer renders a (tailored) RenderCV master object to PDF by
// writing it back to YAML and shelling out to the `rendercv` CLI. Sits
// alongside HtmlPdfRenderer: the JSON-Resume path uses html/template +
// chromedp, this path uses the user's real rendercv config so the Typst
// theme/design is preserved exactly. Mirrors rendercv-renderer.ts.
type RenderCvRenderer struct {
	outDir string
	bin    string
}

func NewRenderCvRenderer(outDir, bin string) *RenderCvRenderer {
	if outDir == "" {
		outDir = "/data/documents"
	}
	if bin == "" {
		bin = "rendercv"
	}
	return &RenderCvRenderer{outDir: outDir, bin: bin}
}

// Render writes master to <outDir>/<baseName>.yaml and runs
// `rendercv render <yaml> -o <outDir> -pdf <baseName>.pdf -nopng -nohtml -nomd`,
// returning both file paths.
func (r *RenderCvRenderer) Render(ctx context.Context, master RendercvMaster, baseName string) (yamlPath, pdfPath string, err error) {
	if err := os.MkdirAll(r.outDir, 0o755); err != nil {
		return "", "", fmt.Errorf("rendercv: mkdir: %w", err)
	}
	yamlPath = filepath.Join(r.outDir, baseName+".yaml")
	pdfPath = filepath.Join(r.outDir, baseName+".pdf")

	data, err := yaml.Marshal(map[string]any(master))
	if err != nil {
		return "", "", err
	}
	if err := os.WriteFile(yamlPath, data, 0o644); err != nil {
		return "", "", fmt.Errorf("rendercv: write yaml: %w", err)
	}

	cmdCtx, cancel := context.WithTimeout(ctx, 120*time.Second)
	defer cancel()
	cmd := exec.CommandContext(cmdCtx, r.bin, "render", yamlPath, "-o", r.outDir, "-pdf", baseName+".pdf", "-nopng", "-nohtml", "-nomd")
	cmd.Dir = r.outDir
	var stderr bytes.Buffer
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return "", "", fmt.Errorf("rendercv: render failed: %w: %s", err, stderr.String())
	}
	if stderr.Len() > 0 {
		slog.Debug("rendercv stderr", "output", stderr.String())
	}
	return yamlPath, pdfPath, nil
}
